hnn_core/parallel_backends.py

```diff
@@ -0,0 +1,265 @@
+
+        neuron_net = _neuron_network(net.params)
+        dpl = _simulate_single_trial(neuron_net)
+
+        spikedata = neuron_net.get_data_from_neuron()
+
+        return dpl, spikedata
+
+    def simulate(self, net):
+        """Simulate the HNN model
+
+        Parameters
+        ----------
+        net : Network object
+            The Network object specifying how cells are
+            connected.
+
+        Returns
+        -------
+        dpl: list of Dipole
+            The Dipole results from each simulation trial
+        """
+
+        n_trials = net.params['N_trials']
+        dpls = []
+
+        parallel, myfunc = self._parallel_func(self._clone_and_simulate)
+        data = parallel(myfunc(net, idx) for idx in range(n_trials))
+
+        # the assignments below need to be made after any forking
+        for idx in range(n_trials):
+            dpls.append(data[idx][0])
+            spikedata = data[idx][1]
+            net.spiketimes.append(spikedata[0])
+            net.spikegids.append(spikedata[1])
+            net.gid_dict = spikedata[2]  # only have one gid_dict
+
+        return dpls
+
+
+class MPI_backend(object):
+    """The MPI_backend class.
+
+    Parameters
+    ----------
+    n_procs : int | None
+        The number of processors (cores) to start MPI job over
+    mpi_cmd : str
+        The name of the mpi launcher executable. Will use 'mpiexec'
+        (openmpi) by default.
+
+    Attributes
+    ----------
+
+    n_jobs : int
+        The number of jobs to start in parallel (NOT SUPPORTED)
+    n_cores : int
+        The number of cores used by the backend
+    mpi_cmd_str : str
+        The string of the mpi command with number of cores and options
+
+    """
+    def __init__(self, n_jobs=1, n_procs=None, mpi_cmd='mpiexec'):
+        import psutil
+        import multiprocessing
+        import os
+        import hnn_core
+
+        try:
+            import mpi4py
+            mpi4py.__version__  # for flake8 test
+        except ImportError:
+            warn('mpi4py not installed. will run on single processor')
+            self.n_procs = 1
+
+        self._type = 'mpi'
+        self.n_jobs = n_jobs
+
+        physical_cores = psutil.cpu_count(logical=False)
+        logical_cores = multiprocessing.cpu_count()
+
+        oversubscribe = False
+        # trying to run on more than available logical cores?
+        if n_procs is None:
+            self.n_cores = logical_cores
+        else:
+            self.n_cores = n_procs
+            if n_procs > logical_cores:
+                oversubscribe = True
+
+        # obey limits set by scheduler
+        try:
+            scheduler_cores = len(os.sched_getaffinity(0))
+        except AttributeError:
+            scheduler_cores = None
+
+        if scheduler_cores is not None:
+            self.n_cores = min(self.n_cores, scheduler_cores)
+
+        # use hwthread-cpus if necessary
+        hyperthreading = False
```

Comment:
is this meant to be exposed to the user? why do you need this variable if not?

Reply:
It's used to build the MPI command string
https://github.com/jonescompneurolab/hnn-core/blob/1cc993d6289ceb30b317a86813f4190a8205fddf/hnn_core/parallel_backends.py#L165-L169

https://github.com/jonescompneurolab/hnn-core/blob/1cc993d6289ceb30b317a86813f4190a8205fddf/hnn_core/parallel_backends.py#L189-L190